File: 39-combination-sum/combination-sum.cpp
class Solution {
public:
    void solve(int i, int target, vector<int>& temp, vector<int>& candidates, vector<vector<int>>& ans) {
        if (target == 0) {
            ans.push_back(temp);
            return;
        }
        if (i == candidates.size() || target < 0) return;

        // Include the current element
        temp.push_back(candidates[i]);
        solve(i, target - candidates[i], temp, candidates, ans);  // we pass `i` (not `i+1`) to allow repetition
        temp.pop_back();  // backtrack

        // Exclude the current element
        solve(i + 1, target, temp, candidates, ans);
    }

    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int> temp;
        solve(0, target, temp, candidates, ans);
        return ans;
    }
};

File: 39-combination-sum/combination-sum.java
class Solution {
    public void solve(int i,int sum,int n,int[] candidates,int target,List<Integer>temp,List<List<Integer>>ans){
        if(i>=n){
            if(sum==target){
                ans.add(new ArrayList<>(temp));
            }
            return;
        }
       // Include current number (if it doesn't exceed target)
        if (sum + candidates[i] <= target) {
            temp.add(candidates[i]);
            // stay at same i to allow reuse
            solve(i, sum + candidates[i], n, candidates, target, temp, ans);
            temp.remove(temp.size() - 1);
        }

        solve(i+1,sum,n,candidates,target,temp,ans);
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        int n=candidates.length;
        List<List<Integer>>ans=new ArrayList<>();
        List<Integer>temp=new ArrayList<>();
        solve(0,0,n,candidates,target,temp,ans);
        return ans;
    }
}